### ISSUE TITLE
[ffi] `Pointer.asTypedList` shared across isolates causes use after free

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#55800]
The native finalizers in `asTypedList` are bound to an isolate, not an isolate group:

https://github.com/dart-lang/sdk/blob/4dd6ee661db6f5b5666a39abf6898360f9f556fa/sdk/lib/ffi/ffi.dart#L368-L373

https://github.com/dart-lang/sdk/blob/4dd6ee661db6f5b5666a39abf6898360f9f556fa/sdk/lib/_internal/vm/lib/ffi_native_finalizer_patch.dart#L115

Hypothesis: The TypedData we create out of Pointer is marked unmodifiable.

We need to either mark the typed data as mutable so that the view does not consider the typed data as unmodifiable, and we copy instead of share the object.
Or, we need to attach finalizers in the isolate group instead of the isolate.

Context:

* https://github.com/dart-lang/native/issues/1151

TODO: verify hypothesis. (Filing issue so that I don't forget.)

**Comment by HosseinYousefi** [comment]
Can we prioritize this?

**Comment by dcharkes** [comment]
TL;DR: **Attaching `NativeFinalizer`s to objects not implementing `Finalizable` breaks assumptions in `Isolate.exit`**.

I've tracked down two issues:

### 1. Isolate.exit shares TypedData without sharing attached NativeFinalizer

`Isolate.exit(sendPort, typedList)` uses a validation pass to see if the objects can be sent to the other isolate. This validation pass prevents `Finalizable`s from being in the object graph, but it does not prevent `ExternalTypedData`s from being in the object graph. No objects are ever copied on this path.

`sendPort.send(typedList)` does not have issues because it will copy the external typed data. And after that the isolate will shut down and the original typed data is freed and its finalizer runs.

We could potentially fix this by (A) during the finalization pass finding the `NativeFinalizer` from the `TypedData` and migrating it to the target isolate near `runtime/lib/isolate.cc::MessageValidator`. We'd also need to instantiate a new native finalizer and add it to the root set of the target isolate.

(We need to make sure we can do this in a reasonable time complexity from the VM. The VM does not have access to the relevant finalizer entries attached to an object, so it would have to loop through _all_ entries in the VM. A possible better approach would be to have the validate function fill a list of typed-datas, and then in the run-finalizers-eager method consume that list and pass in a target isolate.)

Alternatively, we could fix it by (B) changing the finalizer to be a finalizers from the `dart_api.h` instead (as originally proposed in https://github.com/dart-lang/sdk/issues/50507#issuecomment-1529713031). This would technically be a breaking change, see next point.

### 2. We eagerly run `NativeFinalizers` before the `Isolate.exit` to guarantee finalizers have been run

https://github.com/dart-lang/sdk/blob/e8c8500477bd17d9abee2cf2a36a0b1e3414fa67/runtime/vm/isolate.cc#L2619-L2632

For (A) we would need to change those semantics to native finalizers are run eagerly, except for the ones attached to objects being send on exit.

For (B) we don't need to change anything.

### Solutions

A. Migrating `NativeFinalizer`s on exit. (However, we decided against it in https://github.com/dart-lang/sdk/issues/55050 for the general case.)
B. Use isolate-group C API finalizers.

Option B is technically a breaking change, as someone could be using `asTypedList` with a finalizer, and relying on the fact that if they _don't_ send the object to another isolate with `Isolate.exit` the native finalizer is guaranteed to run.

Proposal:

1. Change the semantics of `asTypedList` finalizers to be isolate group scoped.
2. Breaking change announcement.
3. (In some future: make them isolate scoped with https://github.com/dart-lang/sdk/issues/55062)

cc @mraleph @mkustermann

**Comment by mraleph** [comment]
> TL;DR: Attaching NativeFinalizers to objects not implementing Finalizable breaks assumptions in Isolate.exit.

We should actually tighten typing of `NativeFinalizer` implementation because you can currently bypass `Finalizable` requirement by doing a dynamic call. I also see that there are comments in `_NativeFinalizer` which are not correct, e.g. [this one](https://github.com/dart-lang/sdk/blob/1cb6dd08ea358ef93d70a0c8b8c052c4d14ac587/sdk/lib/_internal/vm/lib/ffi_native_finalizer_patch.dart#L67-L68). 

> We could potentially fix this by (A) during the finalization pass finding the `NativeFinalizer` from the `TypedData` and migrating it to the target isolate near `runtime/lib/isolate.cc::MessageValidator`. We'd also need to instantiate a new native finalizer and add it to the root set of the target isolate.

You _could_ probably do that reasonably fast by collecting all encountered external `TypedData` in `MessageValidator` and then checking against these when running native finalizers... So it would not add too much. However it adds unpleasant amount of complexity. 

I think the change you propose is okay. `asTypedList` does not promise finalization on isolate shutdown - so we don't break any promises.

**Comment by dcharkes** [comment]
I realized we implemented this already before, but did not land it: https://dart-review.googlesource.com/c/sdk/+/229544/11. I think we can reuse that implementation with a few modifications. (Rough sketch, I might overlook something.)

* `_detach` in Dart is an `Expando` on the detach key to finalizer ID (unchanged)
  * We omit pruning this map, assuming users will not use some long lived object that outlives the `Finalizable` as `detachkey`.
  * The `detach` method needs to do a runtime entry to cancel the native finalizer
* We don't need `_attached`, because we're running native functions as finalizers, not Dart functions.
  * So, we also don't need the port.
* The `ActiveFinalizers` in the VM maps finalizer IDs to finalizable handles.
  * This data structure has locking in place because it's VM global. (It would not require locking if it were isolate-local, the GC cannot run concurrently with isolate code. It would require locking it was isolate-group local due to possible concurrency between isolates.)
  * Use `@Native` leaf calls instead of native entries (faster)
  * The attach, detach, and finalizer runs modify this data structure.
* Note: `FinalizerIds` are used to avoid having `FinalizableHandle*` as unique identifiers, because those addresses can be reused.
